monai/data/dataset.py

```diff
@@ -850,21 +850,20 @@ def _load_cache_item(self, idx: int):
         return item
 
     def _transform(self, index: int):
-        index_: Any = index
         if self.hash_as_key:
             key = self.hash_func(self.data[index])
-            if key in self._cache:
+            if key in self._hash_keys:
                 # if existing in cache, get the index
-                index_ = key  # if using hash as cache keys, set the key
+                index = self._hash_keys.index(key)
 
-        if isinstance(index_, int) and index_ % len(self) >= self.cache_num:  # support negative index
+        if index % len(self) >= self.cache_num:  # support negative index
```

Comment:
this is a bit confusing. at this point the  variable "index" becomes  an index within cached array (potentially much shorter than the full self.data), we should probably call it something else "index_hash" to avoid later confusion with an actual index in self.data.  And retrieve the cache here
right away data = self._cache[index_hash]

Reply:
I agree the logic here is a little bit confusing.
Same as above comment, if the target item is cached, so we update the index to the hashed list, in this case, it definitely will not call `super()._transform(index)`. That's 2 different cases: fetch `cached item` vs `uncached item`.

Thanks.

Reply:
changing index  ( to hash list space) can lead to  future bugs that is difficult to detect, if in the future  someone else edits this class  they will need to remember that "index" could be in different spaces. you can use a new variable "index_"  (or return cache sooner).  Also the way we detect if it is  in the cache is also confusing `if index % len(self) >= self.cache_num`,  we rely on self.cache_num to be correctly set and index to be in the correct (non-hashed space).  I think more explicit check will be more robust

```
if index < len(self._cache):
  data = self._cache[index]
 else:
  return super()._transform(index) 
```

Reply:
Sure, I updated all this part logic in the latest commit to make it more straightforward.
And `self.cache_num` should always represent the length of cache content, we may use a larger buffer as `self._cache` in future.
thanks for your suggestion.